java/arcs/core/analysis/EntityTypeUnion.kt

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.analysis
+
+import arcs.core.data.EntityType
+import arcs.core.data.FieldName
+import arcs.core.data.FieldType
+import arcs.core.data.Schema
+import arcs.core.data.SchemaFields
+
+/**
+ * Computes the union of the two [Schema] instances. Returns [Outcome.Failure] if the union
+ * is not possible as the inputs are incompatible.
+ */
+fun Schema.union(other: Schema): Outcome<Schema> {
+    val newNames = names.union(other.names)
+    val newFields = fields.union(other.fields).getOrElse { return Outcome.Failure(it) }
+    // TODO(bgogul): hash, refinement, query
+    return Schema(names = newNames, fields = newFields, hash = "").toSuccess()
+}
+
+fun EntityType.union(other: EntityType): Outcome<EntityType> {
+    val newSchema = entitySchema.union(other.entitySchema)
+        .getOrElse { return Outcome.Failure(it) }
+    return EntityType(newSchema).toSuccess()
+}
+
+/**
+ * Returns the result of combining two different field maps.
+ *
+ * If the maps have common [FieldName] entries, the union succeeds if and only if the corresponding
+ * [FieldType] values are the same.
+ */
+private fun Map<FieldName, FieldType>.unionFields(
+    other: Map<FieldName, FieldType>
+): Outcome<Map<FieldName, FieldType>> {
+    val result = mutableMapOf<FieldName, FieldType>()
+    result.putAll(this)
+    other.forEach { entry ->
+        val existing = this.get(entry.key)
+        if (existing != null && entry.value != existing) {
+            return Outcome.Failure(
+                "Incompatible types for field '${entry.key}': ${entry.value} vs. $existing."
+            )
+        }
+        result.put(entry.key, entry.value)
+    }
+    return result.toSuccess()
+}
+
+/**
+ * Computes the union of [SchemaField] instances. Returns [Outcome.Failure] if the union
+ * results in any incompatibility. e.g., incompatible [FieldType] with the same name.
+ */
+private fun SchemaFields.union(other: SchemaFields): Outcome<SchemaFields> {
+    val newSingletons = singletons.unionFields(other.singletons)
+        .getOrElse { return Outcome.Failure(it) }
```

Comment:
It occurs to me now that this `getOrElse` function is not that useful in its current form. A new helper function for this would be nice. e.g. `unwrapOr` (or you could reuse `getOrElse`, but change its lambda to accept the `Outcome.Failure` rather than just the failure message.

I'm imagining something like this:

```
fun Outcome.unwrapOr(block: (Outcome.Failure) -> Unit): T = when(this) {
  is Success -> value
  is Failure -> block(this)
}
```

Then you could use it like so:

```
a.union(b).unwrapOr { return it }
```

Which is a bit shorter.

(The typing of the lambda lets you easily get at the error message, if you still want to do that: `x.unwrapOr { it.reason }`)

Reply:
I was also not happy with the `getOrElse` API when I started using it. Let me address these in a separate PR?

Reply:
Separate PR SGTM. 

I don't really understand the type parameter issue you're describing. Does this not work?

```
inline fun getOrElse(onFailure: (Outcome<T>.Failure) -> T): T = when(this) {
    is Success -> value
    is Failure -> onFailure(this)
}

val x = Outcome.Failure<Int>("error")
x.getOrElse { 7 }
x.getOrElse { throw IllegalArgumentException(it.reason) }
x.getOrElse { return it }
```

Reply:
`getOrElse` is inline. Therefore `return it` exits the function. 

```
fun foo(): Outcome<R> {
  val x = Outcome<T>.getOrElse { 
     // type mismatch as `it` is Outcome<T>, but foo() returns Outcome<R>
      return it 
   }
}
```

Reply:
Ahh, thanks, now I understand what you meant. That's a shame :-( 

I can't think of anything elegant...

Reply:
:(

Reply:
This is giving me strong "Option/Either" monad vibes, e.g. Either<Success, Failure> and getOrElse = fmap :)

Reply:
Precisely. :) I went down the route of `getOrElse` as it was analogous to Kotlin's `Result` type.